core/src/subgraph/runner.rs

```diff
@@ -1041,7 +1102,22 @@ where
 
         self.revert_state(subgraph_ptr.number)?;
 
-        Ok(Action::Continue)
+        // If any datasource has reached its end block in the range [revert_to_ptr, subgraph_ptr]
+        // then we need to restart the blockstream to reset the TriggerFilter.
+        let needs_restart: bool = self
```

Comment:
I think this should be a function to improve readability

Reply:
Yeah the logic is same but the context in which `has_expired` is used is different.